src/django_components/context.py

```diff
@@ -38,77 +81,82 @@ def set_slot_fill(context: Context, component_id: str, slot_name: str, value: "F
-    We pass through the root context to allow configure how slot fills should be rendered.
-
-    See the `SLOT_CONTEXT_BEHAVIOR` setting.
+    See `set_outer_root_context` for more details.
     """
-    return context.get(_OUTER_CONTEXT_CONTEXT_KEY)
+    return context.get(_OUTER_ROOT_CTX_CONTEXT_KEY)
 
 
-def set_root_context(context: Context, root_ctx: Context) -> None:
+def set_outer_root_context(context: Context, outer_ctx: Optional[Context]) -> None:
     """
-    Use this function to set the root context.
-
-    Root context is the top-most context, AKA the context that was passed to
-    the initial `Template.render()`.
-    We pass through the root context to allow configure how slot fills should be rendered.
+    Use this function to set the outer root context.
 
-    See the `SLOT_CONTEXT_BEHAVIOR` setting.
-    """
-    context.push({_OUTER_CONTEXT_CONTEXT_KEY: root_ctx})
+    When we consider a component's template, then outer context is the context
+    that was available just outside of the component's template (AKA it was in
+    the PARENT template).
 
+    Once we have the outer context, next we get the outer ROOT context. This is
+    the context that was available at the top level of the PARENT template.
 
-def capture_root_context(context: Context) -> None:
+    We pass through this context to allow to configure how slot fills should be
+    rendered using the `SLOT_CONTEXT_BEHAVIOR` setting.
     """
-    Set the root context if it was not set before.
-
-    Root context is the top-most context, AKA the context that was passed to
-    the initial `Template.render()`.
-    We pass through the root context to allow configure how slot fills should be rendered.
-
-    See the `SLOT_CONTEXT_BEHAVIOR` setting.
-    """
-    root_ctx_already_defined = _OUTER_CONTEXT_CONTEXT_KEY in context
-    if not root_ctx_already_defined:
-        set_root_context(context, copy(context))
-
-
-def set_slot_component_association(context: Context, slot_id: str, component_id: str) -> None:
+    if outer_ctx and len(outer_ctx.dicts) > 1:
+        outer_root_context: Context = outer_ctx.new()
+        # NOTE_1:
+        # - Index 0 are the defaults set in BaseContext
+        # - Index 1 is the context generated by `Component.get_context_data`
+        #   of the parent's component
+        # - All later indices (2, 3, ...) are extra layers added by the rendering
+        #   logic (each Node usually adds it's own context layer)
+        outer_root_context.push(outer_ctx.dicts[1])
+    else:
+        outer_root_context = Context()
+
+    # Include the mappings.
+    if _SLOT_COMPONENT_ASSOC_KEY in context:
+        outer_root_context[_SLOT_COMPONENT_ASSOC_KEY] = context[_SLOT_COMPONENT_ASSOC_KEY]
+    if _FILLED_SLOTS_CONTENT_CONTEXT_KEY in context:
+        outer_root_context[_FILLED_SLOTS_CONTENT_CONTEXT_KEY] = context[_FILLED_SLOTS_CONTENT_CONTEXT_KEY]
+
+    context[_OUTER_ROOT_CTX_CONTEXT_KEY] = outer_root_context
```

Comment:
Tbh even though I wrote it, the assignments in the `context.py` file feel a bit arcane to me - in other words I haven't wrapped my head comfortably around when to use `Context.push`, when to use `Context[key]`, etc.

...But as things are now it works, and trying to delete any of the parts resulted in errors.